src/algos/fastrp.rs

```diff
@@ -0,0 +1,313 @@
+//! Fast Random Projection
+//!
+//! This method works by:
+//!   1) constructing a node similarity matrix
+//!   2) very sparse random projection for dimension reduction
+
+use nalgebra::DMatrix;
+use nalgebra_sparse::csr::CsrMatrix;
+use rand::prelude::*;
+use rand::Rng;
+use rand_distr::Binomial;
+use nalgebra_sparse::convert::serial::convert_csr_dense;
+use rand_xorshift::XorShiftRng;
+
+use crate::embeddings::EmbeddingStore;
+use crate::embeddings::Distance;
+use crate::progress::CLProgressBar;
+use std::fmt::Write;
+
+/// Fast Random Projection
+///
+/// Compute an embedding per-row of A.
+/// * `A` - the adjacency matrix in CSR format
+/// * `dims` - dimension of embeddings
+/// * `weights` - factor used for weighting of transition matrix power k
+/// * `norm_powers` - if true L2-normalizes matrix powers
+/// * `seed` - a seed for the random number generator
+pub fn get_fastrp_embeddings(
+    A: &CsrMatrix<f32>,
+    dims: usize,
+    weights: &[f32],
+    norm_powers: bool,
+    seed: u64,
+) -> EmbeddingStore {
+    let mut rng = XorShiftRng::seed_from_u64(seed);
+
+    let embeddings = fast_random_projection(
+        A, dims, weights, norm_powers, &mut rng
+    );
+
+    let mut es = EmbeddingStore::new(A.nrows(), dims, Distance::Cosine);
+    embeddings.row_iter()
+    .into_iter()
+    .enumerate()
+    .for_each(|(i,row)| {
+        es.set_embedding(i, row.clone_owned().data.as_vec());
+    });
+    return es;
+}
+
+// Computes a list of powers of the transition matrix of A
+fn fast_random_projection(
+    A: &CsrMatrix<f32>,
+    dim: usize,
+    weights: &[f32],
+    norm_powers: bool,
+    rng: &mut impl Rng
+) -> DMatrix<f32> {
+
+    let k = weights.len();
+
+    // Compute normalizer matrix (inverse degree matrix)
+    let mut a_sum = csr_row_sum(&A);
+    a_sum.iter_mut().for_each(|x| *x = 1.0 / *x); // FIXME: there is a possibility for division by zero here
```

Comment:
@Refefer wasn't sure how to deal with this. In the case that a row of the matrix sums to zero, this will lead to division by zero. Not sure how to resolve that problem.